config/deploy.rb

```diff
@@ -89,6 +102,40 @@ def main_deploy_server
   end
 end
 
+task :install_node do
+  on roles(:app) do
+    with rails_env: fetch(:rails_env) do
+      begin
+        execute fetch(:fnm_install_node_command)
+      rescue SSHKit::Command::Failed
+        begin
+          execute fetch(:fnm_setup_command)
+        rescue SSHKit::Command::Failed
+          execute fetch(:fnm_install_command)
+        else
+          execute fetch(:fnm_update_command)
+        end
+
+        execute fetch(:fnm_install_node_command)
+      end
+    end
+  end
+end
+
+task :map_node_bins do
```

Comment:
Hi @javierm, I'm experiencing an error when deploying in a server prepared with the new version of the installer (one without the former nodejs version 10):

```
00:39 delayed_job:restart
      01 /home/deploy/consul/rvm1scripts/rvm-auto.sh . bundle exec bin/delayed_job -m -n 2 restart
      01 bundler: failed to load command: bin/delayed_job (bin/delayed_job)
      01
      01 /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/execjs-2.8.1/lib/execjs/runtimes.rb:58:in `autodetect'
      01 : Could not find a JavaScript runtime. See https://github.com/rails/execjs for a list of available runtimes. (ExecJS::RuntimeUnavailable)
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/execjs-2.8.1/lib/execjs.rb:5:in `<module:ExecJS>'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/execjs-2.8.1/lib/execjs.rb:4:in `<top (required)>'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/activesupport-6.1.7.4/lib/active_support/dependencies.rb:332:in `require'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/activesupport-6.1.7.4/lib/active_support/dependencies.rb:332:in `block in require'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/activesupport-6.1.7.4/lib/active_support/dependencies.rb:299:in `load_dependency'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/activesupport-6.1.7.4/lib/active_support/dependencies.rb:332:in `require'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/autoprefixer-rails-8.2.0/lib/autoprefixer-rails/processor.rb:2:in `<top (required)>'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/autoprefixer-rails-8.2.0/lib/autoprefixer-rails.rb:36:in `require_relative'
      01 	from /home/deploy/consul/shared/bundle/ruby/3.1.0/gems/autoprefixer-rails-8.2.0/lib/autoprefixer-rails.rb:36:in `<top (required)>'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:60:in `require'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:60:in `block (2 levels) in require'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:55:in `each'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:55:in `block in require'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:44:in `each'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/runtime.rb:44:in `require'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler.rb:186:in `require'
      01 	from /home/deploy/consul/releases/20231010154502/config/application.rb:20:in `<top (required)>'
      01 	from /home/deploy/consul/releases/20231010154502/config/environment.rb:2:in `require_relative'
      01 	from /home/deploy/consul/releases/20231010154502/config/environment.rb:2:in `<top (required)>'
      01 	from bin/delayed_job:3:in `require'
      01 	from bin/delayed_job:3:in `<top (required)>'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli/exec.rb:58:in `load'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli/exec.rb:58:in `kernel_load'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli/exec.rb:23:in `run'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli.rb:486:in `exec'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/vendor/thor/lib/thor/command.rb:27:in `run'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/vendor/thor/lib/thor/invocation.rb:127:in `invoke_command'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/vendor/thor/lib/thor.rb:392:in `dispatch'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli.rb:31:in `dispatch'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/vendor/thor/lib/thor/base.rb:485:in `start'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/cli.rb:25:in `start'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/gems/bundler-2.3.26/libexec/bundle:48:in `block in <top (required)>'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/3.1.0/bundler/friendly_errors.rb:120:in `with_friendly_errors'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/lib/ruby/gems/3.1.0/gems/bundler-2.3.26/libexec/bundle:36:in `<top (required)>'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/bin/bundle:31:in `load'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/bin/bundle:31:in `<main>'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/bin/ruby_executable_hooks:22:in `eval'
      01 	from /home/deploy/.rvm/rubies/ruby-3.1.4/bin/ruby_executable_hooks:22:in `<main>'
```


It seems it fails because we are not initializing the nodejs version with fnm. I tried to add delayed_job to map_node_bins with no luck.

Also realized that we probably need to install nodejs for the other application roles in case they run in separated servers: `web app db importer cron background`. What do you think?

Reply:
> Also realized that we probably need to install nodejs for the other application roles in case they run in separated servers: web app db importer cron background. What do you think?

Right now, we only install Ruby on the `:app` role, so if we change this, I guess we'd have to change that too :thinking:.

I don't know about this because I've never worked in an environment where a server didn't have the `app` role.

Reply:
> I tried to add delayed_job to map_node_bins with no luck

Just to confirm, have you tried adding `bin/delayed_job`, `delayed_job`, or both?

Reply:
> I tried to add delayed_job to map_node_bins with no luck

Just to confirm, have you tried adding `bin/delayed_job`, `delayed_job`, or both?

Maybe we could try setting `:delayed_job_bin_path` to `fnm exec bin` :thinking:, although it looks like a hack.

Reply:
I tried both `delayed_job`, `bin/delayed_job`. Today I tried `bundle exec bin/delayed job` but it failed because the command contains multiple words.

Now I tried `bundle`, and it works fine. Two deployment commands were changed after this update:

```
00:38 whenever:update_crontab
      01 export PATH="$HOME/.fnm:$PATH" && cd /home/deploy/consul/releases/20231011070838 && eval "$(fnm env)" && fnm exec /home/deploy/consul/rvm1scripts/rvm-auto.sh . bundle exec whe…
      01 [write] crontab file updated
      01
    ✔ 01 deploy@127.0.0.1 1.106s
00:40 delayed_job:restart
      01 export PATH="$HOME/.fnm:$PATH" && cd /home/deploy/consul/releases/20231011070838 && eval "$(fnm env)" && fnm exec /home/deploy/consul/rvm1scripts/rvm-auto.sh . bundle exec bin…
      01 delayed_job.0: trying to stop process with pid 45804 sending TERM and waiting 20s ...
      01 delayed_job.0: process with pid 45804 successfully stopped.
      01
      01 delayed_job.0: process with pid 231965 started.
      01
      01 delayed_job.1: trying to stop process with pid 45822 sending TERM and waiting 20s ...
      01
      01 delayed_job.1: process with pid 45822 successfully stopped.
      01
      01 delayed_job.1: process with pid 231979 started.
    ✔ 01 deploy@127.0.0.1 12.065s
```

Reply:
> Right now, we only install Ruby on the :app role, so if we change this, I guess we'd have to change that too 🤔.

> I don't know about this because I've never worked in an environment where a server didn't have the app role.

If we want to run a separate server for the background tasks (delayed_job) we can set the `app` and the `background` roles to that server and it should work. 👍🏼 

I'm also not an expert in executing consul in separate servers with different roles. I was worried that if we wanted to restart delayed_job with Capistrano in a server that does not have the nodejs version it would fail. But the solution is as easy as adding the role app to that server and that's it, capistrano will install Ruby, nodejs and all the dependencies.

Reply:
> Now I tried bundle, and it works fine. Two deployment commands were changed after this update:

Great catch!

While testing this changed, I noticed two additional issues we had:

* Running tasks like `cap staging deploy:restart` wouldn't work because the `map_node_bins` task wasn't executed in this case
* Running `cap staging puma:start` failed as well (although `puma:restart` worked)

So I've updated the code in order to make sure `map_node_bins` is always executed and `puma` is included in the `fnm_map_bins` list.

Reply:
So now, instead of running `rake`, we run:

```bash
export PATH="$HOME/.fnm:$PATH" && cd /home/deploy/consul/current && eval "$(fnm env)" && fnm exec /home/deploy/consul/rvm1scripts/rvm-auto.sh . bundle exec rake
```

Sometimes the world is a horrible place :rofl:.

Reply:
> Sometimes the world of programming is a horrible place 🤣.

Yes, I had the same feeling these days 😓 . Hopefully, at some point we might simplify it a bit 🤞🏼

Reply:
> So I've updated the code in order to make sure map_node_bins is always executed and puma is included in the fnm_map_bins list. [Changes done today](https://github.com/consuldemocracy/consuldemocracy/compare/588c43892ca..d6eb6ce6ad3353049d826600d8e60e6815c83f80).

It seems everything works like a charm now! 👍🏼